Add warning icon on media that lacks descriptive text

dariusk opened the pull request:
This PR adds a warning icon to media that appears in user timelines without descriptive text. This is a signal to people who do not want to boost media that lacks descriptive text. It also is a signal to people who post media -- they will see it in their own timeline with the warning icon, which prompts them to add descriptive text if they have forgotten to do so.

What follows is examples of how it renders for each media type, and then a description of a custom CSS feature, and then a list of known issues and concerns related to this PR.

### Images

Gallery:

![image](https://user-images.githubusercontent.com/266454/209991671-e4c3e665-6285-4e03-9e45-812cca3e4582.png)

Single image:

![image](https://user-images.githubusercontent.com/266454/209991707-6fd54151-a0b2-4145-b6a3-530023a12347.png)

Gallery in mobile view:

![image](https://user-images.githubusercontent.com/266454/209994979-d8e1f481-6227-4c16-b7a8-e5c49b418aa3.png)

### gifv

![image](https://user-images.githubusercontent.com/266454/209991748-12e68a9a-0ca1-4d3d-b823-9e6465a2be1e.png)

### Audio

![image](https://user-images.githubusercontent.com/266454/209991784-4616edc2-003f-40eb-8311-afa5a4e7b443.png)

### Video

![image](https://user-images.githubusercontent.com/266454/209991824-b1304be0-4203-42cd-aa02-dea3f0b0ab39.png)

### Cursor behavior

The cursor, when it floats over the icon, remains as `cursor: default` to indicate that it is not an actionable button, and the tooltip text says in English, "No descriptive text was provided for this media."

### Custom CSS

Also, for admins who want to add custom CSS rules via the admin panel to media that is missing descriptions, there is now a `.media-missing-description` class that wraps all the above media types, so admins can further customize the behavior. Example:

```css
.media-missing-description: {
  border: 2px dashed red;
  box-styling: border-box;
}
```

The icon itself may also be customized via `.media__no-description-icon`.

![image](https://user-images.githubusercontent.com/266454/209992025-f78cf08c-e149-4235-a80d-f3f96915c31c.png)

### Known issues

~There is an alignment issue in a gallery where the left two icons appear 4px from the left but the right two icons appear 6px from the left. I think this is due to the gap between images factoring in to the right two icons, but I haven't been able to fix it yet:~

__Fixed in 0864a9526a__

### Concerns

There may be some wish for all icons on a media item to be the same size. For example, the "GIF" icon on a gifv is a different size from the warning icon. It was already a non-standard size but it's more evident when it is next to another icon.

~Also, I think it's unfortunate that the icon lives on the lower right in video/audio but on the lower left in image/gifv. Maybe it makes sense to move the icon to the lower right? For whatever reason though, seeing the "eye-slash" icon in the upper left and the "warning" icon in the lower right on an image looks weirder to me than having them both left-justified:~

__Update, I have moved the icons to the lower right for images and gifv__:

![image](https://user-images.githubusercontent.com/266454/210055405-9e992467-4513-4943-b3ec-59a1f4e35283.png)

Finally, I predict some users wanting to turn this off. I did not provide an off switch for this in Hometown on the user level, though admins can provide custom CSS to turn the icons off:

```css
.media__no-description-icon {
  display: none;
}
```

I strongly believe that this is a good feature for a healthy network, which is why I do not provide user opt-out here. But if someone on the Mastodon project wants to add it, of course I can't stop you -- but I won't do that work.

dariusk committed:
Add icon on media that lacks alt description

A warning icon now appears on media that lacks an alt description.

Also, for admins who want to add custom CSS rules to media that is
missing descriptions, there is now a `.media-missing-description` rule
that can be added to the custom CSS settings.

Commit:
Add media missing description icon button class

For custom CSS. Also renaming one of the other related classes.

trwnh commented:
Would it not be better to instead reverse this so it shows an "ALT" badge on images with alt text? I'm concerned that the warning triangle exclamation is not at all clear what it is supposed to be warning about. See #18394 for a description and mockups / prior art on the "ALT" badge.

dariusk commented:
@trwnh I strongly believe that media that lacks alt text should be the media that looks weird and slightly off. Media that has alt text should look beautiful and pleasing.

Of course the warning icon could be changed, I have no issue with that. I just picked the best one I could find in the fontawesome 4.7 icon set that Mastodon uses.

nightpool commented:
Tumblr and Twitter have both added "Alt text indicators" recently and while I think they're useful for discoverability (I can't tell you how many times people reblog something I post with alt text to tumblr with an additional "image description" that's often way too verbose and completely obviated by the alt text I already wrote), I agree that they can get pretty annoying and clutter-y, which is why I think this is a better solution (it puts the friction in the right place).

IIRC, Twitter made their alt text feature opt-in. Tumblr did not. Not sure which model makes the most sense here—obviously users could add custom Stylus rules, but some users use platforms that don't provide for that kind of customization (for example, mobile Chrome and Safari)

nightpool commented:
One compromise option would be to make it opt-in for most posts but always displayed on your own posts, so that you always get the notification for the content you control but don't always need to get it for random posts you're browsing.

emceeaich commented:
> One compromise option would be to make it opt-in for most posts but always displayed on your own posts, so that you always get the notification for the content you control but don't always need to get it for random posts you're browsing.

I think the public call out for making inaccessible posts is necessary.

trwnh commented:
>  the warning icon could be changed, I have no issue with that. I just picked the best one I could find in the fontawesome 4.7 icon set that Mastodon uses.

hm, maybe slash out one of the following?

- comment-o
- commenting-o
- sticky-note-o
- the literal text "ALT"
- align-left or align-right

alternatively you could maybe use one of these without the slash:

- question-circle-o
- low-vision (although this could get confusing with the "hide sensitive image" button currently)

trwnh commented:
in a discussion with a friend, i was suggested another option is to use an X instead of a slash. there is prior art, i'm not sure where from though:

![image](https://user-images.githubusercontent.com/10606431/210021309-b210dab0-5f69-4e90-b453-c42903b07d9a.png)
![image](https://user-images.githubusercontent.com/10606431/210021314-8f190f86-35d5-4ed4-82aa-c335ad546408.png)

although what's interesting here as well is that there are icons for both cases where alt text is present and also where it is missing

dariusk commented:
@trwnh Not sure I think any of the suggested alternatives are better than the warning sign. For reference, here are the suggestions:

`question-circle-o` - probably my favorite of the alternatives

![image](https://user-images.githubusercontent.com/266454/210032116-2b2b8ffb-342a-4fc6-85c3-1b7a832d882b.png)

`low-vision` -- I think way too close to the sensitive media button

![image](https://user-images.githubusercontent.com/266454/210032139-16263c80-afd2-4154-b80e-a7e741e05038.png)

And then with slashes (forgive the crappy mockups):

`comment-o` - I think this would be read as "you cannot comment on this"

![image](https://user-images.githubusercontent.com/266454/210032692-958e5f36-a941-42aa-9805-c124c8db9d80.png)

`commenting-o` - same, "you cannot comment on this"

![image](https://user-images.githubusercontent.com/266454/210032717-141ee7b3-4473-4657-b702-b5ad482ba581.png)

`sticky-note-o`

![image](https://user-images.githubusercontent.com/266454/210032729-6340dbf9-2477-4b4d-8ebd-d25583c870fa.png)

`align-left` - my second-favorite alternative. I think we would have to switch between left and right depending on whether we are in an LTR or RTL layout mode

![image](https://user-images.githubusercontent.com/266454/210032766-170de644-97ba-485a-89b9-967c635b0a7f.png)

`align-right`

![image](https://user-images.githubusercontent.com/266454/210032772-0c346f87-aafb-4865-bf95-e924c5070f2e.png)

As for the literal text "ALT" -- imo this is problematic for localization in a way that "GIF" is not.

I like the warning sign best, question mark second best, slashed-out left/right text third best.

trwnh commented:
@dariusk hm, after seeing those actually rendered, i'd say i prefer slashed align-left/align-right ("no text"), with a secondary preference for slashed sticky-note-o ("no note")

i also think that having the icon in the lower right is better than the lower left, as it prevents the awkward-looking case with `gifv`, and it's closer to where you'd have to look in the `video` case

Commit:
Fix alignment on gallery items lacking alt text

The icons now are in the lower right and they correct for their relative inset values.

dariusk commented:
@trwnh I moved the icons to the lower right and also fixed the alignment bug.

![image](https://user-images.githubusercontent.com/266454/210055315-73531825-3f2c-44df-ba5a-1f3bc51990c8.png)

Commit:
Fixing a PropType

hs4man21 commented:
Re: ALT badge vs. warning icon discussion, I think the warning icon is a good approach, but I'm not sure which is preferable. I made a PR some time ago with an ALT badge (https://github.com/mastodon/mastodon/pull/22472) and the iOS app repo already has a similar ALT badge to Twitter incorporated in dev. It seems the Mastodon team made designs for that one in Figma, so a discussion with them is probably needed. I'm happy to update my PR to match those designs if the badge is preferred, but I like the icon, too. Maybe there's a middle ground, like the warning icon in feeds and then the option for sighted users to view the alt text by clicking the image to expand it and then having an ALT button accessible there, since alt text can be useful to sighted users, too. But I don't know. Hopefully this context is useful!

trwnh commented:
more prior art: Akkoma's frontend uses an "align right" kind of icon to represent their toggle for showing image descriptions

![image](https://user-images.githubusercontent.com/10606431/210110642-79a8ce71-7ead-4ffc-8b20-1dbd4f9a3434.png)

dariusk commented:
@hs4man21 Good point, I don't really keep up with the app designs. Not sure how to bring that work into here, though.

dariusk commented:
@trwnh I guess the align-text icons are a pretty good compromise overall.

ClearlyClaire commented on app/javascript/mastodon/components/media_gallery.js:
```diff
@@ -162,12 +164,13 @@ class Item extends React.PureComponent {
           target='_blank'
           rel='noopener noreferrer'
         >
+          { !attachment.get('description') && <IconButton className='media-gallery__item-no-description media__no-description-icon' title={this.props.noDescriptionTitle} icon='exclamation-triangle' style={{ right: `calc(4px + ${(left === 'auto' ? '0px' : left)})`, bottom: `calc(4px + ${(top === 'auto' ? '0px' : top)})` }} overlay /> }
```
I'm not super happy with hardcoding style values like this, but it's consistent with what's done in that class overall, so I'm fine with it.

ClearlyClaire commented on app/javascript/mastodon/components/media_gallery.js:
```diff
@@ -24,6 +25,7 @@ class Item extends React.PureComponent {
     displayWidth: PropTypes.number,
     visible: PropTypes.bool.isRequired,
     autoplay: PropTypes.bool,
+    noDescriptionTitle: PropTypes.string,
```
Not entirely happy with passing the description down as a prop since it will be the same in all cases. I guess it spares a few `intl` calls though.

ClearlyClaire commented on app/javascript/mastodon/features/audio/index.js:
```diff
@@ -555,6 +556,7 @@ class Audio extends React.PureComponent {
             </div>
 
             <div className='video-player__buttons right'>
+              {!alt && <button type='button' title={intl.formatMessage(messages.no_descriptive_text)} aria-label={intl.formatMessage(messages.no_descriptive_text)} className='player-button no-action media__no-description-icon' ><Icon id='exclamation-triangle' fixedWidth /></button>}
```
Using a `button` for something that has no action sounds slightly off to me, but I imagine that using anything else might cause subtle alignment issues with neighboring elements which *are* buttons.

ClearlyClaire reviewed (approved):
Overall, I'm happy with this change, though I understand the warning icon might not immediately convey the appropriate meaning. Out of the suggested alternatives, I do not like `question-circle-o` that much because it looks like it could actually be used for the alt-text itself, `align-left` and `align-right` could work, but I'm not super confident with adding the slash ourselves… though we could make a custom icon like we did for the disabled boost icon.

trwnh commented:
> I strongly believe that media that lacks alt text should be the media that looks weird and slightly off. Media that has alt text should look beautiful and pleasing.

Looks like #24782 just got merged which adds an ALT label...

dariusk commented:
Well that is disappointing.

mjankowski commented:
Closing due to merge conflicts and the assumption that the latest linked issue makes this no longer relevant -- but if you still want to push this foward can you rebase and re-open?

dariusk commented:
I'm guessing this is dead in the water ideologically, but since I'll be maintaining it in the hometown fork it should be easy enough to rebase and reopen.
